scipy/_lib/uarray/_backend.py

```diff
@@ -0,0 +1,363 @@
+    (1, 'a')
+
+    See Also
+    --------
+    uarray
+        See the module documentation for how to override the method by creating backends.
+    """
+    kw_defaults, arg_defaults, opts = get_defaults(argument_extractor)
+    ua_func = _Function(
+        argument_extractor,
+        argument_replacer,
+        domain,
+        arg_defaults,
+        kw_defaults,
+        default,
+    )
+
+    return functools.update_wrapper(ua_func, argument_extractor)
+
+
+def set_backend(backend, *args, **kwargs):
+    """
+    A context manager that sets the preferred backend. Uses :obj:`BackendOptions` to create
+    the options, and sets the backend with the preferred options.
+
+    Parameters
+    ----------
+    backend
+        The backend to set.
+
+    See Also
+    --------
+    BackendOptions: The backend plus options.
+    skip_backend: A context manager that allows skipping of backends.
+    """
+    return _SetBackendContext(backend, *args, **kwargs)
+
+
+def skip_backend(backend):
+    """
+    A context manager that allows one to skip a given backend from processing
+    entirely. This allows one to use another backend's code in a library that
+    is also a consumer of the same backend.
+
+    Parameters
+    ----------
+    backend
+        The backend to skip.
+
+    See Also
+    --------
+    set_backend: A context manager that allows setting of backends.
+    """
+    return _SkipBackendContext(backend)
+
+
+def get_defaults(f):
+    sig = inspect.signature(f)
+    kw_defaults = {}
+    arg_defaults = []
+    opts = set()
+    for k, v in sig.parameters.items():
+        if v.default is not inspect.Parameter.empty:
+            kw_defaults[k] = v.default
+        if v.kind in (
+            inspect.Parameter.POSITIONAL_ONLY,
+            inspect.Parameter.POSITIONAL_OR_KEYWORD,
+        ):
+            arg_defaults.append(v.default)
+        opts.add(k)
+
+    return kw_defaults, tuple(arg_defaults), opts
+
+
+def set_global_backend(backend):
+    """
+    This utility method replaces the default backend for permanent use. It
+    will be tried in the list of backends automatically, unless the
+    ``only`` flag is set on a backend. This will be the first tried
+    backend outside the :obj:`set_backend` context manager.
+
+    Note that this method is not thread-safe.
+
+    .. warning::
+        We caution library authors against using this function in
+        their code. We do *not* support this use-case. This function
+        is meant to be used only by users themselves, or by a reference
+        implementation, if one exists.
+
+    Parameters
+    ----------
+    backend
+        The backend to register.
+    """
+    _uarray.set_global_backend(backend)
+
+
+def register_backend(backend):
+    """
+    This utility method sets registers backend for permanent use. It
+    will be tried in the list of backends automatically, unless the
+    ``only`` flag is set on a backend.
+
+    Note that this method is not thread-safe.
```

Comment:
It would be good to specify which backend takes precedence when multiple are registered (both here and in the uarray docs).
    
Looking at the uarray C++ code, it would seem the backends are checked in the order in which they were registered?

Reply:
I believe the docs used to say that the order wasn't guarunteed. I guess that was removed in the doc rewrite. cc @hameerabbasi

Reply:
The order is as follows:

* The context managers are tried first, innermost to outermost.
* The global backend is tried.
* The registered backends are tried.

If at any point an `only` or `coerce` is encountered, the process stops.

Reply:
There is no guarantee for the order of the registered backends, the rest is guaranteed.

Reply:
@hameerabbasi: right, but if `register_backend` is called twice, is it guaranteed that the registered backends are tried in the order they were registered?

Reply:
okay, maybe just add a comment to the docstring mentioning something like: 

If `register_backend` is called more than once, the order in which the registered backends are tried is not guaranteed